src/log.rs

```diff
@@ -64,35 +64,88 @@ macro_rules! emit_event {
 pub trait LogExt<T> {
     /// Emits a warning if the receiver contains an Err value.
     ///
-    /// Returns an [`Option<T>`] with the `Ok(_)` value, if any:
-    /// - You won't get any warnings about unused results but can still use the value if you need it
-    /// - This prevents the same warning from being printed to the log multiple times
-    ///
     /// Thanks to the [track_caller](https://blog.rust-lang.org/2020/08/27/Rust-1.46.0.html#track_caller)
     /// feature, the location of the caller is printed to the log, just like with the warn!() macro.
+    ///
+    /// Unfortunately, the track_caller feature does not work on async functions (as of Rust 1.50).
+    /// Once it is, you can add `#[track_caller]` to helper functions that use `log()` and `log_m()`
+    /// so that the location of the caller can be seen in the log. (this won't work with the macros,
+    /// like warn!(), since the file!() and line!() macros don't work with track_caller)  
+    /// See https://github.com/rust-lang/rust/issues/78840 for progress on this.
+    #[track_caller]
+    fn log_err(self, context: &Context) -> Self;
+
+    /// Emits a warning if the receiver contains an Err value.
+    ///
+    /// Unlike `log_err()`, returns an [`Option<T>`] instead of Self.
+    ///
+    /// Example:
+    /// ```text
+    /// if let Err(e) = do_something() {
+    ///     warn!(context, "{:#}", e);
+    /// }
+    /// ```
+    /// is equivalent to:
+    /// ```text
+    /// do_something().ok_or_log(context);
+    /// ```
+    /// and is also equivalent to:
+    /// ```text
+    /// do_something().log_err(context).ok();
+    /// ```
+    ///
+    /// For a note on the `track_caller` feature, see the doc comment on `log_err()`.
     #[track_caller]
-    fn log(self, context: &Context) -> Option<T>;
+    fn ok_or_log(self, context: &Context) -> Option<T>;
+
+    /// Like `ok_or_log()`, but you can pass an extra message that is prepended in the log.
+    ///
+    /// Example:
+    /// ```text
+    /// if let Err(e) = do_something() {
+    ///     warn!(context, "Something went wrong: {:#}", e);
+    /// }
+    /// ```
+    /// is equivalent to:
+    /// ```text
+    /// do_something().ok_or_log_msg(context, "Something went wrong");
+    /// ```
+    /// and is also equivalent to:
+    /// ```text
+    /// do_something().context("Something went wrong").ok_or_log(context);
+    /// ```
+    ///
+    /// For a note on the `track_caller` feature, see the doc comment on `log_err()`.
+    #[track_caller]
+    fn ok_or_log_msg(self, context: &Context, msg: &'static str) -> Option<T>;
 }
 
 impl<T> LogExt<T> for anyhow::Result<T> {
```

Comment:
oh, why is this only for `anyhow::Result`?  Doesn't a more general impl work on just Result?  The ffi version does `impl<T: Default, E: std::fmt::Display> ResultExt<T, E> for Result<T, E> {` I think you should be able to do something similar here.
You could make `ok_or_log()` a provided method and impl `log_err` & `ok_or_log_msg` for both anyhow and a more generic result.  You could also change `ok_or_log_msg` not to rely on anyhow's context and also make it a provided method.  If you want to go even more DRY you can add a `format_log()` and also provide `log_err()` so that only `format_log()` needs to be specialised for anyhow.

Reply:
and once you're doing all that you can probably use this LogExt trait in the FFI instead of the existing ResultExt.  but i don't mind if you leave the ffi out of this PR (though you already started modifying it)